src/sympc/protocol/falcon/falcon.py

```diff
@@ -451,3 +454,102 @@ def select_shares(x: MPCTensor, y: MPCTensor, b: MPCTensor) -> MPCTensor:
         z = x + (d * (y - x))
 
         return z
+
+    @staticmethod
+    def _random_prime_group(
+        session: Session, shape: Union[torch.Size, tuple]
+    ) -> MPCTensor:
+        """Computes shares of random number in Zp*.Zp* is the multiplicative group mod p.
+
+        Args:
+            session (Session): session to generate random shares for.
+            shape (Union[torch.Size, tuple]): shape of the random share to generate.
+
+        Returns:
+            share (MPCTensor): Returns shares of random number in group Zp*.
+
+        Zp* = {1,2..,p-1},where p is a prime number.
+        We use Euler's Theorem for verifying that random share is not zero.
+        It states that:
+        For a general modulus n
+        a^phi(n) = 1(mod n), if a is co prime to n.
+        In our case n=p(prime number), phi(p) = p-1
+        phi(n) = Euler totient function.
+        We effectively try to sample a random number in range [1,p-1],discard the instances where
+        it equals zero.
+        """
+        while True:
+            ptr_list: List[ReplicatedSharedTensor] = []
+            for session_ptr in session.session_ptrs:
+                ptr = session_ptr.prrs_generate_random_share(
+                    shape=(), ring_size=str(PRIME_NUMBER)
+                ).resolve_pointer_type()
+                ptr = ptr.repeat(shape)
+                ptr_list.append(ptr)
+
+            m = MPCTensor(shares=ptr_list, session=session, shape=shape)
+
+            m_euler = m ** (PRIME_NUMBER - 1)
+
+            if (m_euler.reconstruct(decode=False) == 1).all():
+                return m
+
+    @staticmethod
+    def private_compare(x: List[MPCTensor], r: torch.Tensor) -> MPCTensor:
```

Comment:
Q: Should we do any check to make sure that the `r` has at least `ring_size` bits?

Reply:
By ring_size you mean the session tensor type geoge

Reply:
Currently we place the onus on the user to provide the correct tensor type.

Reply:
Including a sanity check for r would be good :+1:  , I would include it.

Reply:
I think we do not need a sanity check, in retrospection, it works for arbitrary tensor types also , as we do an internal type conversion in RSTensor apply_op,

Reply:
If we are comparing 64 bit number with  8 bit r values, the positions at [9,63] would be zero, which yields the correct result

Reply:
Ahh...it is fine. I was thinking that it is `undefined behaviour` for shifting (to left/right) more than the number of bits for that data type.